### PR TITLE
Prevent unnecessary map copying

### PR DESCRIPTION
**Pull request by idegtiarenko**
image

According to the image above, the majority of time spend in ClusterInfoSimulator is related to copying immutable map and creating a hashmap. This change remove that for 2 maps that are not changing during the calculation.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java [hunk 1]
```diff
@@ -27,8 +27,8 @@ public ClusterInfoSimulator(ClusterInfo clusterInfo) {
         this.leastAvailableSpaceUsage = new HashMap<>(clusterInfo.getNodeLeastAvailableDiskUsages());
         this.mostAvailableSpaceUsage = new HashMap<>(clusterInfo.getNodeMostAvailableDiskUsages());
         this.shardSizes = new HashMap<>(clusterInfo.shardSizes);
-        this.shardDataSetSizes = new HashMap<>(clusterInfo.shardDataSetSizes);
-        this.dataPath = new HashMap<>(clusterInfo.dataPath);
+        this.shardDataSetSizes = clusterInfo.shardDataSetSizes;
+        this.dataPath = clusterInfo.dataPath;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java:31]: Those 2 are immutable maps at this point. They would be reused by Map.copyOf. This should help a lot as shardDataSetSizes size is proportional to a shard count.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java:31]: I think I'd prefer to use `Map.copyOf()` here just to be explicit that we're not modifying anything in `clusterInfo`. (we could reasonably `assert this.shardDataSetSizes == clusterInfo.shardDataSetSizes` to protect against a performance regression)

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-1

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-2

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-1

**Pull request merged**